arch/arm/src/samv7/sam_lin_sock.c

```diff
@@ -0,0 +1,1503 @@
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+static void sam_lin_reset(struct sam_lin_s *priv)
+{
+  ninfo("LIN%" PRIu8 "\n", priv->port);
+
+  /* Reset and disable receiver and transmitter */
+
+  sam_lin_putreg(priv, SAM_UART_CR_OFFSET, UART_CR_RSTRX | UART_CR_RSTTX |
+                                           UART_CR_RXDIS | UART_CR_TXDIS);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: sam_linsockinitialize
+ *
+ * Description:
+ *   Initialize the selected LIN port as LIN socket interface
+ *
+ * Input Parameters:
+ *   Port number (for hardware that has multiple LIN interfaces)
+ *
+ * Returned Value:
+ *   OK on success; Negated errno on failure.
+ *
+ ****************************************************************************/
+
+int sam_linsockinitialize(int port)
+{
+  struct sam_lin_s *priv = NULL;
+
+  ninfo("LIN%" PRIu8 "\n", port);
+
+  /* Select the LIN device structure */
+
+  switch (port)
+    {
+#ifdef CONFIG_SAMV7_USART0_LIN_SOCKET
+      case 0:
+        priv = &g_lin0priv;
+
+        sam_usart0_enableclk();
+
+        sam_configgpio(GPIO_USART0_RXD);
+        sam_configgpio(GPIO_USART0_TXD);
+        break;
+#endif
+
+#ifdef CONFIG_SAMV7_USART1_LIN_SOCKET
+      case 1:
+        priv = &g_lin1priv;
+
+        sam_usart1_enableclk();
+
+        sam_configgpio(GPIO_USART1_RXD);
+        sam_configgpio(GPIO_USART1_TXD);
+        break;
+#endif
+
+#ifdef CONFIG_SAMV7_USART2_LIN_SOCKET
+      case 2:
+        priv = &g_lin2priv;
+
+        sam_usart2_enableclk();
+
+        sam_configgpio(GPIO_USART2_RXD);
+        sam_configgpio(GPIO_USART2_TXD);
+        break;
+#endif
+
+      default:
+        nerr("ERROR: Unsupported port %d\n", port);
+        return -ENODEV;
+    }
+
+  /* Setup device private information */
+
+  priv->dev.d_private = priv;
+
+  /* Put the interface in the down state.  This usually amounts to resetting
+   * the device and/or calling sam_lin_ifdown().
+   */
+
+  ninfo("callbacks done\n");
+
+  sam_lin_ifdown(&priv->dev);
+
+  /* Register the device with the OS so that socket IOCTLs can be performed */
+
+  return netdev_register(&priv->dev, NET_LL_CAN);
+}
+
+/****************************************************************************
+ * Name: arm_netinitialize
+ *
+ * Description:
+ *   Initialize the LIN device interfaces.  If there is more than one device
+ *   interface in the chip, then board-specific logic will have to provide
+ *   this function to determine which, if any, LIN interfaces should be
+ *   initialized.
+ *
+ ****************************************************************************/
+
+#if !defined(CONFIG_NETDEV_LATEINIT)
+void arm_netinitialize(void)
```

Comment:
Just an idea not directly related to this PR, but perhaps we could have a file `sam_netinitialize.c` that would define `arm_netinitialize` for both LIN and Ethernet. The current state leads to compile error for multiple definitions of `arm_netinitialize` if `CONFIG_NETDEV_LATEINIT` is not set. But I think there is no reason why we could not configure both LIN and Ethernet in `arm_netinitialize`, right?

I can make a PR that would move `arm_netinitialize`  from `sam_ethernet.c` to `sam_netinitialize.c` and then you can move LIN there.

Reply:
I think this is a next step PR. Let's merge this one first